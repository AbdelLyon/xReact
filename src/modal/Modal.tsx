import {
  Modal as NextUIModal,
  ModalContent,
  ModalHeader,
  ModalBody,
  ModalFooter,
  Button,
  useDisclosure,
  ModalProps,
} from "@nextui-org/react";
import { useState } from "react";
import { SlotsToClasses } from "@nextui-org/react";

import { cn } from "@/utils";

type GenericModalProps = {
  trigger: React.ReactNode;
  title?: React.ReactNode;
  footer?: React.ReactNode;
  modalProps?: Partial<ModalProps>;
  children: React.ReactNode;
  onAction?: () => void;
  buttonCloseLabel?: string;
  buttonActionLabel?: string;
  classNames?: SlotsToClasses<
    | "footer"
    | "backdrop"
    | "base"
    | "body"
    | "header"
    | "closeButton"
    | "wrapper"
  >;
};

export const Modal: React.FC<GenericModalProps> = ({
  trigger,
  title = "Modal Title",
  footer,
  modalProps,
  onAction,
  buttonCloseLabel = "Close",
  buttonActionLabel,
  classNames,
  children,
}) => {
  const { isOpen, onOpen, onClose } = useDisclosure();
  const [backdrop, setBackdrop] = useState<ModalProps["backdrop"]>("opaque");

  const handleOpen = (backdropType: ModalProps["backdrop"] = "opaque") => {
    setBackdrop(backdropType);
    onOpen();
  };

  const handleAction = () => {
    onAction?.();
    onClose();
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
    if (event.key === "Enter" || event.key === " ") {
      handleOpen();
    }
  };

  return (
    <>
      <div
        role="button"
        tabIndex={0}
        onClick={() => handleOpen()}
        onKeyDown={handleKeyDown}
      >
        {trigger}
      </div>

      <NextUIModal
        backdrop={backdrop}
        classNames={{
          closeButton: cn("absolute right-4 top-4", classNames?.closeButton),
          base: cn("bg-background", classNames?.base),
          ...classNames,
        }}
        isOpen={isOpen}
        onClose={onClose}
        {...modalProps}
      >
        <ModalContent>
          {(onClose) => (
            <>
              <ModalHeader className="flex flex-col gap-1">{title}</ModalHeader>

              <ModalBody>{children}</ModalBody>

              <ModalFooter>
                {footer || (
                  <>
                    <Button
                      className="border-primary/20"
                      color="primary"
                      radius="sm"
                      variant="bordered"
                      onPress={onClose}
                    >
                      {buttonCloseLabel}
                    </Button>
                    {buttonActionLabel && onAction && (
                      <Button
                        color="primary"
                        radius="sm"
                        onPress={handleAction}
                      >
                        {buttonActionLabel}
                      </Button>
                    )}
                  </>
                )}
              </ModalFooter>
            </>
          )}
        </ModalContent>
      </NextUIModal>
    </>
  );
};
