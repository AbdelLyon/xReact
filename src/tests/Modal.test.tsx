import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";

import { Modal } from "@/modal";

describe("Modal Component", () => {
  it("renders correctly with title, description, trigger, and matches snapshot", () => {
    const { container } = render(
      <Modal
        description="This is a test description"
        title="Test Modal Title"
        trigger={<button>Open Modal</button>}
      >
        <div>Modal Content</div>
      </Modal>,
    );

    expect(container).toMatchSnapshot();
  });

  it("renders with custom footer and matches snapshot", () => {
    const { container } = render(
      <Modal
        footer={<button>Confirm</button>}
        title="Modal with Footer"
        trigger={<button>Open Modal</button>}
      >
        <div>Modal Content</div>
      </Modal>,
    );

    expect(container).toMatchSnapshot();
  });

  it("renders without description and matches snapshot", () => {
    const { container } = render(
      <Modal
        title="Modal without Description"
        trigger={<button>Open Modal</button>}
      >
        <div>Modal Content</div>
      </Modal>,
    );

    expect(container).toMatchSnapshot();
  });
});
