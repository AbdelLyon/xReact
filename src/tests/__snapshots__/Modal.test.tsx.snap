// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal Component > renders correctly with title, description, trigger, and matches snapshot 1`] = `
<div>
  <button
    aria-controls="radix-:r0:"
    aria-expanded="false"
    aria-haspopup="dialog"
    data-state="closed"
    type="button"
  >
    Open Modal
  </button>
</div>
`;

exports[`Modal Component > renders with custom footer and matches snapshot 1`] = `
<div>
  <button
    aria-controls="radix-:r3:"
    aria-expanded="false"
    aria-haspopup="dialog"
    data-state="closed"
    type="button"
  >
    Open Modal
  </button>
</div>
`;

exports[`Modal Component > renders without description and matches snapshot 1`] = `
<div>
  <button
    aria-controls="radix-:r6:"
    aria-expanded="false"
    aria-haspopup="dialog"
    data-state="closed"
    type="button"
  >
    Open Modal
  </button>
</div>
`;
